xdsl/irdl/constraints.py

```diff
@@ -380,15 +379,14 @@ def verify(
     ) -> None:
         constraint_context = constraint_context or ConstraintContext()
         for attr_constr in self.attr_constrs:
-            # Copy the constraint to ensure that if the constraint fails, the
-            # constraint context is not modified.
-            constraint_context_copy = constraint_context.copy()
+            # Push a new scope that will be discarded if the constraint fails to verify,
+            # and kept if verification succeeds
+            constraint_context.push_variable_scope()
```

Comment:
Could we make this a contextmanager?

Reply:
I tried that, and the scope has to stay once the thing succeeds, so it's not as symmetric as the context manager needs

Reply:
I'd forgotten about that point. In that case this seems reasonable. The only other thing worth considering is that if an AnyOf succeeds then the scope could be "flattened" onto the one below. I imagine this would almost never improve performance though.

Reply:
I think eventually we might want to implement a more advanced data structure with checkpointing that can do this more efficiently by default, but I don't think there's a rush